File: app/src/main/java/com/atrainingtracker/trainingtracker/fragments/aftermath/ChangeSportAndEquipmentDialogFragment.java
package com.atrainingtracker.trainingtracker.fragments.aftermath;

import android.app.Dialog;
import android.content.ContentValues;
import android.content.DialogInterface;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import com.atrainingtracker.R;
import com.atrainingtracker.banalservice.BSportType;
import com.atrainingtracker.banalservice.database.SportTypeDatabaseManager;
import com.atrainingtracker.trainingtracker.database.EquipmentDbHelper;
import com.atrainingtracker.trainingtracker.database.WorkoutSummariesDatabaseManager;
import com.atrainingtracker.trainingtracker.database.WorkoutSummariesDatabaseManager.WorkoutSummaries;

import java.util.List;

public class ChangeSportAndEquipmentDialogFragment extends DialogFragment {

    private static final String TAG = "ChangeSportAndEquipmentDialogFragment";
    private static final boolean DEBUG = true;

    // ARGUMENTS
    private static final String ARG_WORKOUT_ID = "workoutId";
    private static final String ARG_CURRENT_SPORT_ID = "currentSportId";
    private static final String ARG_CURRENT_EQUIPMENT_NAME = "currentEquipmentName";

    // VIEWS
    private Spinner mSpinnerSport;
    private Spinner mSpinnerEquipment;
    private TextView mTvEquipmentLabel;


    // DATA
    private long mWorkoutId;
    private long mCurrentSportId;
    private List<Long> mSportTypeIdList;
    private String mCurrentEquipmentName;

    public interface OnSportChangedListener {
        void onSportChanged(long workoutId);
    }
    public void setOnSportChangedListener(OnSportChangedListener listener) {
        mListener = listener;
    }

    private OnSportChangedListener mListener;

    public static ChangeSportAndEquipmentDialogFragment newInstance(long workoutId, long currentSportId, String currentEquipmentName) {
        ChangeSportAndEquipmentDialogFragment fragment = new ChangeSportAndEquipmentDialogFragment();
        Bundle args = new Bundle();
        args.putLong(ARG_WORKOUT_ID, workoutId);
        args.putLong(ARG_CURRENT_SPORT_ID, currentSportId);
        args.putString(ARG_CURRENT_EQUIPMENT_NAME, currentEquipmentName);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mWorkoutId = getArguments().getLong(ARG_WORKOUT_ID);
            mCurrentSportId = getArguments().getLong(ARG_CURRENT_SPORT_ID);
            mCurrentEquipmentName = getArguments().getString(ARG_CURRENT_EQUIPMENT_NAME);
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        LayoutInflater inflater = requireActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.dialog_change_sport, null);

        mSpinnerSport = view.findViewById(R.id.spinner_change_sport);
        mSpinnerEquipment = view.findViewById(R.id.spinner_change_equipment);
        mTvEquipmentLabel = view.findViewById(R.id.tv_equipment_label);

        setupSportSpinnerAdapter();
        setInitialSportSelection();
        setupEquipmentSpinner();

        builder.setView(view)
                .setTitle(R.string.change_sport)
                .setPositiveButton(R.string.save, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int id) {
                        saveNewSportAndEquipment();
                    }
                })
                .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        ChangeSportAndEquipmentDialogFragment.this.getDialog().cancel();
                    }
                });
        return builder.create();
    }

    /**
     * This is adapted directly from EditWorkoutFragment.
     * We simplify it to always show all sport types.
     */
    private void setupSportSpinnerAdapter() {
        if (DEBUG) Log.i(TAG, "setSpinnerSport");

        List<String> sportTypeUiNameList = SportTypeDatabaseManager.getSportTypesUiNameList();
        mSportTypeIdList = SportTypeDatabaseManager.getSportTypesIdList();

        mSpinnerSport.setAdapter(new ArrayAdapter<>(getActivity(), android.R.layout.simple_spinner_item, sportTypeUiNameList.toArray(new String[0])));
    }

    private void setInitialSportSelection() {
        int selectionIndex = mSportTypeIdList.indexOf(mCurrentSportId);
        if (selectionIndex != -1) {
            // Set selection WITHOUT triggering the onItemSelected listener.
            mSpinnerSport.setSelection(selectionIndex, false);
        }
    }

    private void setupEquipmentSpinner() {
        if (DEBUG) Log.i(TAG, "setSpinnerEquipment");

        updateEquipmentSpinner(mCurrentSportId, mCurrentEquipmentName);

        mSpinnerSport.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                long selectedSportId = mSportTypeIdList.get(position);
                // Update equipment list for the newly selected sport, with no pre-selected equipment
                updateEquipmentSpinner(selectedSportId, null);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }

    /**
     * Updates the equipment spinner's content based on the selected sport.
     * @param currentSportId The ID of the currently selected sport.
     * @param equipmentToSelect The name of the equipment to pre-select, or null.
     */
    private void updateEquipmentSpinner(long currentSportId, @Nullable String equipmentToSelect) {

        // first, get the BSportType
        BSportType bSportType = SportTypeDatabaseManager.getInstance().getBSportType(currentSportId);

        int labelResId = switch (bSportType) {
            case BIKE -> R.string.equipment_type_bike;
            case RUN -> R.string.equipment_type_shoe;
            default -> R.string.Equipment;
        };
        mTvEquipmentLabel.setText(labelResId);

        EquipmentDbHelper equipmentDbHelper = new EquipmentDbHelper(getActivity());
        List<String> equipmentNameList = equipmentDbHelper.getEquipment(bSportType);

        if (equipmentNameList.isEmpty()) {
            mTvEquipmentLabel.setVisibility(View.GONE);
            mSpinnerEquipment.setVisibility(View.GONE);
        } else {
            mTvEquipmentLabel.setVisibility(View.VISIBLE);
            mSpinnerEquipment.setVisibility(View.VISIBLE);

            mSpinnerEquipment.setAdapter(new ArrayAdapter<>(getActivity(), android.R.layout.simple_spinner_item, equipmentNameList.toArray(new String[0])));

            // when we reselect the original sport, the original equipment should be selected
            if (currentSportId == mCurrentSportId) {
                equipmentToSelect = mCurrentEquipmentName;
            }

            if (equipmentToSelect != null) {
                int equipmentSelectionIndex = equipmentNameList.indexOf(equipmentToSelect);
                if (equipmentSelectionIndex != -1) {
                    mSpinnerEquipment.setSelection(equipmentSelectionIndex);
                }
            }
        }
    }

    /**
     * Saves the newly selected sport ID to the database for the given workout.
     */
    private void saveNewSportAndEquipment() {
        // get selected sport
        int selectedPosition = mSpinnerSport.getSelectedItemPosition();
        if (selectedPosition < 0 || selectedPosition >= mSportTypeIdList.size()) {
            Log.e(TAG, "Invalid sport selection");
            return;
        }

        long newSportTypeId = mSportTypeIdList.get(selectedPosition);
        if (DEBUG) Log.d(TAG, "Saving new sportId: " + newSportTypeId + " for workout: " + mWorkoutId);

        // get selected equipment
        String selectedEquipmentName = mSpinnerEquipment.getSelectedItem().toString();
        EquipmentDbHelper equipmentDbHelper = new EquipmentDbHelper(getActivity());
        long newEquipmentId = equipmentDbHelper.getEquipmentId(selectedEquipmentName);
        if (DEBUG) Log.d(TAG, "Saving new equipment: " + selectedEquipmentName + " (id: " + newEquipmentId + ")");

        // save
        ContentValues values = new ContentValues();
        values.put(WorkoutSummaries.SPORT_ID, newSportTypeId);
        values.put(WorkoutSummaries.EQUIPMENT_ID, newEquipmentId);

        SQLiteDatabase db = WorkoutSummariesDatabaseManager.getInstance().getOpenDatabase();
        try {
            db.update(WorkoutSummaries.TABLE,
                    values,
                    WorkoutSummaries.C_ID + "=" + mWorkoutId,
                    null);
        } catch (SQLException e) {
            // TODO: use Toast?
            Log.e(TAG, "Error while writing" + e);
        }

        // After saving, notify the listener that a change was made.
        if (mListener != null) {
            mListener.onSportChanged(mWorkoutId);
        }
    }
}

File: app/src/main/java/com/atrainingtracker/trainingtracker/ui/components/workoutheader/ChangeSportAndEquipmentDialogFragment.kt
package com.atrainingtracker.trainingtracker.ui.components.workoutheader

import android.app.Dialog
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.core.view.isVisible
import androidx.fragment.app.DialogFragment
import com.atrainingtracker.R
import com.atrainingtracker.banalservice.BSportType
import com.atrainingtracker.banalservice.database.SportTypeDatabaseManager
import com.atrainingtracker.trainingtracker.database.EquipmentDbHelper

class ChangeSportAndEquipmentDialogFragment : DialogFragment() {

    private val TAG = "ChangeSportAndEquipment"

    // Modern, lambda-based listener that passes back the selected IDs.
    // The fragment is responsible for the database logic.
    var onSave: ((newSportId: Long, newEquipmentId: Long) -> Unit)? = null

    // Arguments are loaded safely and lazily.
    private val initialSportId: Long by lazy { requireArguments().getLong(ARG_CURRENT_SPORT_ID) }
    private val initialEquipmentName: String? by lazy { requireArguments().getString(ARG_CURRENT_EQUIPMENT_NAME) }

    // Views
    private lateinit var sportSpinner: Spinner
    private lateinit var equipmentSpinner: Spinner
    private lateinit var equipmentLabel: TextView

    // Data
    private val sportTypeIds: List<Long> by lazy { SportTypeDatabaseManager.getInstance(requireContext()).getSportTypesIdList() }
    private val sportTypeNames: List<String> by lazy { SportTypeDatabaseManager.getInstance(requireContext()).getSportTypesUiNameList() }
    private lateinit var equipmentDbHelper: EquipmentDbHelper

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val view = requireActivity().layoutInflater.inflate(R.layout.dialog_change_sport, null)
        equipmentDbHelper = EquipmentDbHelper(requireContext())

        // Initialize Views
        sportSpinner = view.findViewById(R.id.spinner_change_sport)
        equipmentSpinner = view.findViewById(R.id.spinner_change_equipment)
        equipmentLabel = view.findViewById(R.id.tv_equipment_label)

        setupSportSpinner()
        setupEquipmentSpinner() // Initial setup for equipment spinner adapter

        // Set initial selections and listeners
        setInitialSportSelection()
        updateEquipmentSpinner(initialSportId, initialEquipmentName)
        setupSportSelectionListener()

        return AlertDialog.Builder(requireContext())
            .setView(view)
            .setTitle(R.string.change_sport)
            .setPositiveButton(R.string.save) { _, _ ->
                handleSave()
            }
            .setNegativeButton(R.string.cancel, null)
            .create()
    }

    private fun setupSportSpinner() {
        sportSpinner.adapter = ArrayAdapter(
            requireContext(),
            android.R.layout.simple_spinner_item,
            sportTypeNames
        )
    }

    private fun setInitialSportSelection() {
        val selectionIndex = sportTypeIds.indexOf(initialSportId)
        if (selectionIndex != -1) {
            // Set selection WITHOUT triggering the onItemSelected listener initially.
            sportSpinner.setSelection(selectionIndex, false)
        }
    }

    private fun setupEquipmentSpinner() {
        // Set up the adapter once with an empty list. It will be populated by updateEquipmentSpinner.
        equipmentSpinner.adapter = ArrayAdapter<String>(
            requireContext(),
            android.R.layout.simple_spinner_item,
            mutableListOf()
        )
    }

    private fun setupSportSelectionListener() {
        sportSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                val selectedSportId = sportTypeIds.getOrElse(position) { return }
                // When a new sport is selected, reset the equipment selection.
                val equipmentToSelect = if (selectedSportId == initialSportId) initialEquipmentName else null
                updateEquipmentSpinner(selectedSportId, equipmentToSelect)
            }
            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }
    }

    private fun updateEquipmentSpinner(sportId: Long, equipmentToSelect: String?) {
        val bSportType = SportTypeDatabaseManager.getInstance(requireContext()).getBSportType(sportId);

        // Set the equipment label based on sport type
        equipmentLabel.setText(
            when (bSportType) {
                BSportType.BIKE -> R.string.equipment_type_bike
                BSportType.RUN -> R.string.equipment_type_shoe
                else -> R.string.Equipment
            }
        )

        val equipmentNames = equipmentDbHelper.getEquipment(bSportType)
        val hasEquipment = equipmentNames.isNotEmpty()

        // Use isVisible for cleaner visibility toggling
        equipmentLabel.isVisible = hasEquipment
        equipmentSpinner.isVisible = hasEquipment

        if (hasEquipment) {
            val adapter = equipmentSpinner.adapter as ArrayAdapter<String>
            adapter.clear()
            adapter.addAll(equipmentNames)
            adapter.notifyDataSetChanged()

            if (equipmentToSelect != null) {
                val equipmentIndex = equipmentNames.indexOf(equipmentToSelect)
                if (equipmentIndex != -1) {
                    equipmentSpinner.setSelection(equipmentIndex)
                }
            }
        }
    }

    private fun handleSave() {
        val selectedSportId = sportTypeIds.getOrElse(sportSpinner.selectedItemPosition) {
            Log.e(TAG, "Invalid sport selection on save.")
            return
        }

        // Only get equipment if the spinner is visible
        val selectedEquipmentId = if (equipmentSpinner.isVisible && equipmentSpinner.selectedItem != null) {
            val selectedEquipmentName = equipmentSpinner.selectedItem.toString()
            equipmentDbHelper.getEquipmentId(selectedEquipmentName)
        } else {
            0L // Or your designated ID for "no equipment"
        }

        Log.d(TAG, "Saving SportID: $selectedSportId, EquipmentID: $selectedEquipmentId")

        // Pass the result back to the caller. The dialog is "dumb" and does not save.
        onSave?.invoke(selectedSportId, selectedEquipmentId)
    }

    companion object {
        private const val ARG_CURRENT_SPORT_ID = "currentSportId"
        private const val ARG_CURRENT_EQUIPMENT_NAME = "currentEquipmentName"

        @JvmStatic
        fun newInstance(currentSportId: Long, currentEquipmentName: String?): ChangeSportAndEquipmentDialogFragment {
            return ChangeSportAndEquipmentDialogFragment().apply {
                arguments = bundleOf(
                    ARG_CURRENT_SPORT_ID to currentSportId,
                    ARG_CURRENT_EQUIPMENT_NAME to currentEquipmentName
                )
            }
        }
    }
}